### PR TITLE
Add option for machinepool to create Spot MachineSets

### PR DESCRIPTION
**Pull request by kahootali**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/Azure/go-autorest/autorest/azure/auth v0.4.1
 	github.com/Azure/go-autorest/autorest/to v0.3.1-0.20191028180845-3492b2aff503
 	github.com/aws/aws-sdk-go v1.30.16
+	github.com/blang/semver v3.5.1+incompatible
```
Comment (staebler) [inline go.mod:11]: Why is this change needed?

File: pkg/awsclient/mock/client_generated.go [hunk 2]
```diff
@@ -1,7 +1,6 @@
 // Code generated by MockGen. DO NOT EDIT.
 // Source: ./client.go
 
-// Package mock is a generated GoMock package.
```
Comment (staebler) [inline pkg/awsclient/mock/client_generated.go:4]: None of the changes to the mock files should be needed.

File: pkg/controller/remotemachineset/awsactuator.go [hunk 3]
```diff
@@ -230,6 +237,22 @@ func decodeAWSMachineProviderSpec(rawExt *runtime.RawExtension, scheme *runtime.
 	return spec, nil
 }
 
+func (a *AWSActuator) updateSpotPrice(ms *machineapi.MachineSet, pool *hivev1.MachinePool) error {
+	if pool.Spec.Platform.AWS.SpotMarketOptions == nil {
+		return nil
+	}
+	awsProviderSpec, err := decodeAWSMachineProviderSpec(ms.Spec.Template.Spec.ProviderSpec.Value, a.scheme)
```
Comment (staebler) [inline pkg/controller/remotemachineset/awsactuator.go:244]: You should not need to decode. The `ms.Spec.Template.Spec.ProviderSpec.Value.Object` field should already be populated.

File: pkg/controller/remotemachineset/awsactuator.go [hunk 4]
```diff
@@ -230,6 +237,22 @@ func decodeAWSMachineProviderSpec(rawExt *runtime.RawExtension, scheme *runtime.
...
+	if err != nil {
+		return err
+	}
+	awsProviderSpec.SpotMarketOptions = &awsprovider.SpotMarketOptions{
```
Comment (staebler) [inline pkg/controller/remotemachineset/awsactuator.go:248]: Can this be done in `updateProviderConfig`?

File: pkg/apis/hive/v1/aws/machinepool.go [hunk 5]
```diff
@@ -16,6 +16,18 @@ type MachinePoolPlatform struct {
 
 	// EC2RootVolume defines the storage for ec2 instance.
 	EC2RootVolume `json:"rootVolume"`
+
+	// SpotMarketOptions allows users to configure instances to be run using AWS Spot instances.
+	*SpotMarketOptions `json:"spotMarketOptions,omitempty"`
```
Comment (staebler) [inline pkg/apis/hive/v1/aws/machinepool.go:21]: Please give an explicit name to this field.

File: pkg/awsclient/mock/client_generated.go (hunk 2 above)
Comment (kahootali) [reply pkg/awsclient/mock/client_generated.go:4]: I had installed latest version of mockgen, I think it converted the files, will remove them

File: go.mod (hunk 1 above)
Comment (kahootali) [reply go.mod:11]: This change was generated automatically when I ran make target, but shouldn't this be part, as it is being used in clusterversion_controller

Comment (staebler) [reply go.mod:11]: You are correct. This line should be present. This change was missed in https://github.com/openshift/hive/pull/1041. I will create a separate PR for this change to keep it isolated from the spot instance changes.

Comment (staebler) [reply go.mod:11]: https://github.com/openshift/hive/pull/1050

File: pkg/apis/hive/v1/aws/machinepool.go (hunk 5 above)
Comment (kahootali) [reply pkg/apis/hive/v1/aws/machinepool.go]: Done

File: pkg/controller/remotemachineset/awsactuator.go (hunk 4 above)
Comment (kahootali) [reply pkg/controller/remotemachineset/awsactuator.go]: Done

File: pkg/controller/remotemachineset/awsactuator.go (hunk 3 above)
Comment (kahootali) [reply pkg/controller/remotemachineset/awsactuator.go]: Done

File: pkg/awsclient/mock/client_generated.go (hunk 2 above)
Comment (kahootali) [reply pkg/awsclient/mock/client_generated.go]: Removed

File: go.mod (hunk 1 above)
Comment (kahootali) [reply go.mod]: Pulled the change

**Comment by kahootali** [comment]
@staebler Can you re-review it please? I have made the changes

File: pkg/apis/hive/v1/aws/machinepool.go [hunk 6]
```diff
@@ -16,6 +16,18 @@ type MachinePoolPlatform struct {
 
 	// EC2RootVolume defines the storage for ec2 instance.
 	EC2RootVolume `json:"rootVolume"`
+
+	// SpotMarketOptions allows users to configure instances to be run using AWS Spot instances.
+	SpotMarketOptions *SpotMarketOptions `json:"spotMarketOptions,omitempty"`
```
Comment (staebler) [inline pkg/apis/hive/v1/aws/machinepool.go:21]:
```suggestion
	// +optional
	SpotMarketOptions *SpotMarketOptions `json:"spotMarketOptions,omitempty"`
```

File: pkg/apis/hive/v1/aws/machinepool.go [hunk 7]
```diff
@@ -16,6 +16,18 @@ type MachinePoolPlatform struct {
...
+}
+
+// SpotMarketOptions defines the options available to a user when configuring
+// Machines to run on Spot instances.
+// Most users should provide an empty struct.
+type SpotMarketOptions struct {
+	// The maximum price the user is willing to pay for their instances
+	// Default: On-Demand price
+	MaxPrice *string `json:"maxPrice,omitempty"`
```
Comment (staebler) [inline pkg/apis/hive/v1/aws/machinepool.go:30]:
```suggestion
	// +optional
	MaxPrice *string `json:"maxPrice,omitempty"`
```

**Comment by kahootali** [comment]
@staebler I did `make update` and pushed but still the `verify` & `e2e` tests are failing, can you check the reason?

**Comment by staebler** [comment]
> @staebler I did `make update` and pushed but still the `verify` & `e2e` tests are failing, can you check the reason?

Run `make update` (or just `make update-codegen`) again. You should get the following changes.
```
$ git diff
diff --git a/pkg/apis/hive/v1/aws/zz_generated.deepcopy.go b/pkg/apis/hive/v1/aws/zz_generated.deepcopy.go
index be122f19..e6ac0880 100644
--- a/pkg/apis/hive/v1/aws/zz_generated.deepcopy.go
+++ b/pkg/apis/hive/v1/aws/zz_generated.deepcopy.go
@@ -34,6 +34,11 @@ func (in *MachinePoolPlatform) DeepCopyInto(out *MachinePoolPlatform) {
                copy(*out, *in)
        }
        out.EC2RootVolume = in.EC2RootVolume
+       if in.SpotMarketOptions != nil {
+               in, out := &in.SpotMarketOptions, &out.SpotMarketOptions
+               *out = new(SpotMarketOptions)
+               (*in).DeepCopyInto(*out)
+       }
        return
 }
 
@@ -70,3 +75,24 @@ func (in *Platform) DeepCopy() *Platform {
        in.DeepCopyInto(out)
        return out
 }
+
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
+func (in *SpotMarketOptions) DeepCopyInto(out *SpotMarketOptions) {
+       *out = *in
+       if in.MaxPrice != nil {
+               in, out := &in.MaxPrice, &out.MaxPrice
+               *out = new(string)
+               **out = **in
+       }
+       return
+}
+
+// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new SpotMarketOptions.
+func (in *SpotMarketOptions) DeepCopy() *SpotMarketOptions {
+       if in == nil {
+               return nil
+       }
+       out := new(SpotMarketOptions)
+       in.DeepCopyInto(out)
+       return out
+}
```

**Comment by kahootali** [comment]
@staebler I tried this but I got no changes, I have go version 1.14.6, so `make update` doesn't work so I ran 
```
make update-codegen GO_REQUIRED_MIN_VERSION:=
hack/update-codegen.sh
Generating deepcopy funcs
Generating clientset for hive:v1 at github.com/openshift/hive/pkg/client/clientset
Generating listers for hive:v1 at github.com/openshift/hive/pkg/client/listers
Generating informers for hive:v1 at github.com/openshift/hive/pkg/client/informers
Generating deepcopy funcs
```
but there were no changes

**Comment by staebler** [comment]
@kahootali You need to have Hive in your GOPATH. The code generators require this. The changes are getting put under your GOPATH.

**Comment by kahootali** [comment]
@staebler I already have Hive in my GOPATH, still cant see any change, ran 

```
make update-codegen GO_REQUIRED_MIN_VERSION:=
make update GO_REQUIRED_MIN_VERSION:=
```
Repo structure
```
Repo Folder: /home/kahoot/go/src/github.com/kahootali/hive
echo $GOPATH
/home/kahoot/go
```

**Comment by staebler** [comment]
@kahootali Hive needs to be at /home/kahoot/go/src/github.com/openshift/hive.

**Comment by kahootali** [comment]
@staebler Squashed & tests are also passing

**Comment by kahootali** [comment]
@staebler done
